# Spotify Data Analysis

Exploratory analysis of a Spotify tracks dataset using Python, Pandas, SQL, and Plotly.

The analysis focuses on artist representation, popularity trends, explicit content across decades, and the relationship between song duration and popularity.

## Project Objective

The goal of this project is to analyze Spotify track data and identify patterns in music popularity, artist representation, explicit content, and song characteristics across different decades.

## Key Questions

- Which artists appear most frequently in the dataset?
- How has music representation changed across decades?
- How has explicit content changed over time?
- Are explicit tracks more popular than non-explicit tracks?
- Is track duration related to popularity?

In [1]:
import pandas as pd
import plotly.express as px

df = pd.read_csv("../data/spotify_analysis.csv")

df.head(10)

,track_name,artist_names,album_release_year,track_duration_ms,explicit,popularity
0,Fader,The Temper Trap,2009,192373,False,0
1,Sherry,Frankie Valli & The Four Seasons,2003,152160,False,54
2,I Took A Pill In Ibiza - Seeb Remix,"Mike Posner, Seeb",2016,197933,True,63
3,Let Go for Tonight,Foxes,2014,238413,False,39
4,The Way I Want To Touch You,Captain & Tennille,1975,163586,False,35
5,Still Got Time (feat. PARTYNEXTDOOR),"ZAYN, PARTYNEXTDOOR",2017,188490,False,53
6,Your Song,Rita Ora,2018,180160,False,60
7,Fix You,Coldplay,2005,295533,False,83
8,There You'll Be - 2007 Remaster,Faith Hill,2007,220840,False,44
9,Every Breath You Take - Remastered 2003,The Police,1983,253886,False,3


## Data Overview

The dataset used in this notebook contains the analysis-ready variables prepared during the data cleaning stage:

- track name
- artist name(s)
- album release year
- track duration
- explicit content flag
- popularity score

## Visual Analysis

### 1. Top 10 Artists by Number of Tracks

In [2]:
top_artists = df["artist_names"].value_counts().head(10).reset_index()

top_artists

,artist_names,count
0,Taylor Swift,50
1,P!nk,47
2,Elvis Presley,46
3,Ed Sheeran,40
4,The Beatles,38
5,U2,38
6,Madonna,35
7,The Rolling Stones,33
8,Katy Perry,30
9,Bee Gees,30


In [3]:
fig = px.bar(
    top_artists,
    x="artist_names",
    y="count",
    title="Top 10 Artists by Number of Tracks",
    labels={
        "artist_names": "Artist",
        "count": "Number of Tracks"
    }
)

fig.show()

**Finding:** Taylor Swift has the highest number of tracks (50), followed by P!nk (47) and Elvis Presley (46), while the top 10 include both modern and classic artists.

### 2. Song Duration vs. Popularity

In [4]:
df["duration_minutes"] = df["track_duration_ms"] / 60000

fig = px.scatter(
    df,
    x="duration_minutes",
    y="popularity",
    title="Song Duration vs. Popularity"
)

fig.show()

In [5]:
correlation = df["duration_minutes"].corr(df["popularity"])

print(f"Correlation between song duration and popularity: {correlation:.3f}")

Correlation between song duration and popularity: 0.029


**Finding:** Most tracks are between 2 and 5 minutes long, with a few outliers. The correlation between duration and popularity is only 0.029, indicating almost no linear relationship between the two.

### 3. Explicit Tracks by Decade

In [6]:
df["decade"] = (df["album_release_year"] // 10) * 10

df[["album_release_year", "decade"]].head(10)

,album_release_year,decade
0,2009,2000
1,2003,2000
2,2016,2010
3,2014,2010
4,1975,1970
5,2017,2010
6,2018,2010
7,2005,2000
8,2007,2000
9,1983,1980


In [7]:
explicit_by_decade = (
    (df.groupby("decade")["explicit"].mean() * 100)
    .reset_index(name="explicit_percentage")
)

explicit_by_decade

,decade,explicit_percentage
0,1950,0.000000
1,1960,0.000000
2,1970,0.279330
3,1980,0.441501
4,1990,3.944316
5,2000,6.096131
6,2010,6.576629
7,2020,13.953488


In [8]:
fig = px.line(
    explicit_by_decade,
    x="decade",
    y="explicit_percentage",
    title="Explicit Tracks by Decade",
    labels={
        "decade": "Decade",
        "explicit_percentage": "Explicit Tracks (%)"
    }
)

fig.show()

**Finding:** The share of explicit tracks has been increasing since the 1980s, with the strongest growth occurring after 2010.

### 4. Average Popularity by Decade

In [9]:
popularity_by_decade = (
    df.groupby("decade")["popularity"]
    .mean()
    .reset_index()
)

popularity_by_decade["popularity"] = popularity_by_decade["popularity"].round(2)

popularity_by_decade

,decade,popularity
0,1950,55.41
1,1960,35.47
2,1970,39.25
3,1980,36.85
4,1990,32.19
5,2000,29.29
6,2010,30.09
7,2020,42.55


In [10]:
fig = px.line(
    popularity_by_decade,
    x="decade",
    y="popularity",
    title="Average Popularity by Decade",
    labels={
        "decade": "Decade",
        "popularity": "Average Popularity"
    }
)

fig.show()

**Finding:** Average popularity generally declined from the 1970s to the 2000s, followed by a strong increase in the 2020s.

### Average Popularity — Explicit vs. Non-Explicit

In [12]:
explicit_popularity = (
    df.groupby("explicit")["popularity"]
    .mean()
    .reset_index()
)

explicit_popularity["explicit"] = explicit_popularity["explicit"].map({
    True: "Explicit",
    False: "Non-Explicit"
})

fig = px.bar(
    explicit_popularity,
    x="explicit",
    y="popularity",
    title="Average Popularity: Explicit vs. Non-Explicit",
    labels={"explicit": "Track Type", "popularity": "Average Popularity"}
)

fig.show()

**Finding:** Explicit tracks have a higher average popularity score (39.85) than non-explicit tracks (32.09) in this dataset.

## Key Findings

- Taylor Swift has the highest number of tracks (50), followed by P!nk (47) and Elvis Presley (46).
- The share of explicit tracks has increased since the 1980s, reaching 13.95% in the 2020s.
- Average popularity generally declined from the 1970s to the 2000s, followed by a strong increase in the 2020s.
- Track duration has almost no linear relationship with popularity (r = 0.029).
- Explicit tracks have a higher average popularity score (39.85) than non-explicit tracks (32.09).